compiler/modules/types/string.bal

```diff
@@ -1,65 +1,151 @@
 // Implementation specific to basic type string.
 
 type StringSubtype readonly & record {|
+    CharStringSubtype char;
+    NonCharStringSubtype nonChar;
+|};
+
+// allowed = false with empty `values` means `String:char`
+type CharStringSubtype readonly & record {|
+    boolean allowed;
+    string:Char[] values;
+|};
+
+// allowed = false with empty `values` mean all non char strings
+type NonCharStringSubtype readonly & record {|
     boolean allowed;
     string[] values;
 |};
 
 public function stringConst(string value) returns SemType {
-    StringSubtype st = { allowed: true, values: [value] };
+    CharStringSubtype char;
+    NonCharStringSubtype nonChar;
+    
+    if value.length() == 1 {
+        char = { allowed: true, values: [<string:Char>value] };
+        nonChar = { allowed: true, values: [] };
+    }
+    else {
+        char = { allowed: true, values: [] };
+        nonChar = { allowed: true, values: [value] };
+    }
+    return uniformSubtype(UT_STRING, { char, nonChar });
+}
+
+function stringChar() returns SemType {
+    StringSubtype st = {
+        char: { allowed: false, values: [] }, 
+        nonChar: { allowed: true, values: [] }
+    };
     return uniformSubtype(UT_STRING, st);
 }
 
 function stringSubtypeSingleValue(SubtypeData d) returns string? {
     if d is boolean {
         return ();
     }
-    StringSubtype s = <StringSubtype>d;
-    if !s.allowed {
-        return ();
+    var { char, nonChar } = <StringSubtype>d;
+    int charCount = char.values.length();
```

Comment:
This still seems overly complicated to me. Can something like this work?

```
int charCount = char.allowed ? char.values.length() : 2;
int nonCharCount = nonChar.allowed ? nonChar.values.length() : 2;
if charCount + nonCharCount == 1 {
    return charCount != 0 ? char.values[0] : nonChar.values[0];
}
```

Reply:
Wow, that's clever.

Reply:
Little bit too clever. We could have written it better, if we had nil lifting `+`.